walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -36,50 +39,171 @@ public int compare(Event a, Event b) {
+    // windows.
+    return optimalMeetingTimes.isEmpty() ? goodWindows : optimalMeetingTimes;
   }
 
-  private Collection<TimeRange> getMeetingTimes(
-      Collection<Event> eventsCollection,
-      MeetingRequest request,
-      boolean includeOptionalAttendees) {
-    HashSet<String> attendees = new HashSet<String>(request.getAttendees());
-    if (includeOptionalAttendees) {
-      attendees.addAll(request.getOptionalAttendees());
+  /**
+   * Returns a mapping of optional attendees to their free times.
+   *
+   * @param events all events to be considered
+   * @param optionalAttendees everyone who needs to attend this meeting
+   * @param minTime minimum length of time for meeting
+   */
+  private HashMap<String, ArrayList<TimeRange>> getOptionalAttendeesFreeTimes(
+      Collection<Event> events, HashSet<String> optionalAttendees, long minTime) {
+    HashMap<String, ArrayList<TimeRange>> times = new HashMap<String, ArrayList<TimeRange>>();
+    for (String attendee : optionalAttendees) {
+      HashSet<String> attendeeSet = new HashSet<String>();
+      attendeeSet.add(attendee);
+
+      // Find all possible meeting times for just this one attendee. Must do this to deal with
+      // double bookings.
+      times.put(attendee, getMandatoryAttendeesMeetingTimes(events, minTime, attendeeSet));
     }
-    ArrayList<Event> events = getRelevantEvents(attendees, new ArrayList<Event>(eventsCollection));
-    List<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
+    return times;
+  }
+
+  /**
+   * Returns a change log of the number of available optional attendees over time. First part of a
+   * sweep-line algorithm.
+   *
+   * @param optionalAttendeesFreeTimes mapping of all attendees to their free times
+   */
+  private ArrayList<Map.Entry<Integer, Integer>> getChanges(
+      HashMap<String, ArrayList<TimeRange>> optionalAttendeesFreeTimes) {
+    TreeMap<Integer, Integer> changes = new TreeMap<Integer, Integer>();
+    for (Map.Entry e : optionalAttendeesFreeTimes.entrySet()) {
+      for (TimeRange time : (ArrayList<TimeRange>) e.getValue()) {
+        changes.put(time.start(), changes.getOrDefault(time.start(), 0) + 1);
+        changes.put(time.end(), changes.getOrDefault(time.end(), 0) - 1);
+      }
+    }
+    return new ArrayList<Map.Entry<Integer, Integer>>(changes.entrySet());
+  }
+
+  /**
+   * Updates optimalMeetingTimes based on current meeting time's attendance.
+   *
+   * @param time current meeting time range
+   * @param bestAttendance best attendance of optional attendees seen so far
+   * @param currAttendance attendance optional attendees of current meeting
+   * @param optionalAttendeesFreeTimes all attendees and their free times
+   * @param minTime minimum meeting time
+   */
+  private int updateOptimalTimes(
+      TimeRange time,
+      int bestAttendance,
+      int currAttendance,
+      ArrayList<TimeRange> optimalMeetingTimes,
```

Comment:
You should try to avoid mutating random arguments (as part of the semantics of a function). It's not obvious to callers when this is happening. Usually you either (1) want to include it in your return values, or (2) use a member variable.

(1) works best if you can create small value classes. We use [AutoValue for that](https://github.com/google/auto/blob/master/value/userguide/index.md), but I won't suggest you bring in that dependency just for this.

For (2), I'm not sure about using a member variable here, because `FindMeetingQuery` might be used for multiple requests, and, if this wasn't a toy example, it could eventually get parameters independent of a single call to `query` at some later time. Conflating state for one call of `query` with state for the class in general is a recipe for hard to maintain code in the future. 

For (2), therefore, what you really probably want is a class representing a single instance of finding a time which can have appropriate member variables. 

No fix requested though, unless you're willing to do what I would do, which is move most of this logic into an inner static class `SingleMeetingResolver`, which has optimalMeetingTimes as a member variable (along with the parameters of query).

To start you off if you want to do that:
```java
  private static class SingleMeetingResolver {
    private final Collection<Event> events;
    private final MeetingRequest request;
    private ArrayList<TimeRange> optimalMeetingTimes = new ArrayList<TimeRange>();
 
    SingleMeetingResolver(Collection<Event> events, MeetingRequest request) {
      this.events = events;
      this.request = request; 
    }
  
    Collection<TimeRange> resolveBestTime() {
      // TODO: Move the existing work in here, but use member variables for most state.
    }
  } 

  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
    return new SingleMeetingResolver(events, request).resolveBestTime();
  } 
```

Reply:
l get what you want me to do, but l don't get the intuition behind it. More specifically this part: "I'm not sure about using a member variable here, because FindMeetingQuery might be used for multiple requests, and, if this wasn't a toy example, it could eventually get parameters independent of a single call to query at some later time. Conflating state for one call of query with state for the class in general is a recipe for hard to maintain code in the future."

Are you thinking of a case where another function outside of FindMeetingQuery calls updateOptimalTimes but doesn't want optimalMeetingTimes modified?